server/src/main/java/org/apache/druid/server/SegmentManager.java

```diff
@@ -171,9 +167,16 @@ public boolean loadSegment(final DataSegment segment) throws SegmentLoadingExcep
           );
 
           if ((entry != null) && (entry.getChunk(segment.getShardSpec().getPartitionNum()) != null)) {
-            log.warn("Told to load a adapter for a segment[%s] that already exists", segment.getId());
+            log.warn("Told to load an adapter for segment[%s] that already exists", segment.getId());
             resultSupplier.set(false);
           } else {
+            final Segment adapter;
+            try {
+              adapter = getAdapter(segment);
```

Comment:
I think this means we will only be able to load 1 segment at a time for a datasource (ConcurrentHashMap.compute synchronizes on the hashtable Node when applying the `remappingFunction`). If so, it's still not parallel enough. Locking at the level of an individual segment would be best; maybe something that involves tracking a Map of currently-loading segments to Objects we can use for synchronization?

Reply:
Oh yeah, I just realized it. I'm thinking about a better way.